File: clients/py/src/features/resolve_package.py
from pathlib import Path
from typing import Any

from polywrap import (
    PolywrapClient,
    PolywrapClientConfigBuilder,
    Uri,
    UriPackage,
    WasmPackage,
    deserialize_wrap_manifest,
)
from pydantic import BaseModel, validator

from validators import validate_root_directory


class TestCaseInput(BaseModel):
    directory: Path
    uri: str

    @validator("directory")
    def valid_root_directory(cls, v: Any):
        return validate_root_directory(
            v, Path(__file__).parent.parent.parent.parent.parent
        )


def run_test_case(input: Any) -> None:
    input_obj = TestCaseInput.parse_obj(input)

    root = Path(__file__).parent.parent.parent.parent.parent
    wrap_dir = root / input_obj.directory
    uri = Uri.from_str(input_obj.uri)

    with open(wrap_dir / "wrap.info", "rb") as f:
        manifest = deserialize_wrap_manifest(f.read())
    with open(wrap_dir / "wrap.wasm", "rb") as f:
        wasm_module = f.read()

    wrap_package = WasmPackage(
        file_reader=NotImplemented, manifest=manifest, wasm_module=wasm_module
    )

    config = PolywrapClientConfigBuilder().set_package(uri, wrap_package).build()

    client = PolywrapClient(config)

    print(f"Resolving URI: {uri}")

    result = client.try_resolve_uri(uri)

    match result:
        case UriPackage():
            print("Received: package")
            print("Success!")
        case _:
            print("Failed!")

File: clients/js/src/features/resolve_package.ts
import { Input } from "../input";

import { PolywrapClientConfigBuilder, PolywrapClient } from "@polywrap/client-js";
import { WasmPackage } from "@polywrap/wasm-js";

import path from "path";
import fs from "fs";

export async function runTestCase(input: unknown): Promise<void> {
  const inputObj = Input.expectObject<{
    directory: unknown;
    uri: unknown;
  }>(input);
  const wrapDir = Input.expectRootDir(
    inputObj.directory,
    path.join(__dirname, "../../../../")
  )
  const uri = Input.expectUri(inputObj.uri).uri;

  const manifest = fs.readFileSync(
    path.join(wrapDir, "wrap.info")
  );
  const wasmModule = fs.readFileSync(
    path.join(wrapDir, "wrap.wasm")
  );
  const wrapPackage = WasmPackage.from(
    manifest,
    wasmModule
  );

  const config = new PolywrapClientConfigBuilder()
    .setPackage(uri, wrapPackage)
    .build();

  const client = new PolywrapClient(config);

  console.log(`Resolving URI: ${uri}`);

  const result = await client.tryResolveUri({ uri });

  if (result.ok) {
    console.log(`Received: ${result.value.type}`);
    console.log("Success!");
  }
}
